src/merkle/smt/mod.rs

```diff
@@ -0,0 +1,383 @@
+
+        let leaf_index: LeafIndex<SMT_DEPTH> = key.into();
+
+        match self.leaves.get_mut(&leaf_index.value()) {
+            Some(leaf) => leaf.insert(key, value),
+            None => {
+                self.leaves.insert(leaf_index.value(), SmtLeaf::Single((key, value)));
+
+                None
+            }
+        }
+    }
+
+    /// Removes key-value pair at leaf index pointed to by `key` if it exists.
+    fn perform_remove(&mut self, key: SmtKey) -> Option<Word> {
+        let leaf_index: LeafIndex<SMT_DEPTH> = key.into();
+
+        if let Some(leaf) = self.leaves.get_mut(&leaf_index.value()) {
+            let (old_value, is_empty) = leaf.remove(key);
+            if is_empty {
+                self.leaves.remove(&leaf_index.value());
+            }
+            old_value
+        } else {
+            // there's nothing stored at the leaf; nothing to update
+            None
+        }
+    }
+}
+
+impl SparseMerkleTree<SMT_DEPTH> for Smt {
+    type Key = SmtKey;
+    type Value = Word;
+    type Leaf = SmtLeaf;
+
+    const EMPTY_VALUE: Self::Value = EMPTY_WORD;
+
+    fn root(&self) -> RpoDigest {
+        self.root
+    }
+
+    fn set_root(&mut self, root: RpoDigest) {
+        self.root = root;
+    }
+
+    fn get_inner_node(&self, index: NodeIndex) -> InnerNode {
+        self.inner_nodes.get(&index).cloned().unwrap_or_else(|| {
+            let node = EmptySubtreeRoots::entry(SMT_DEPTH, index.depth() + 1);
+
+            InnerNode { left: *node, right: *node }
+        })
+    }
+
+    fn insert_inner_node(&mut self, index: NodeIndex, inner_node: InnerNode) {
+        self.inner_nodes.insert(index, inner_node);
+    }
+
+    fn insert_value(&mut self, key: Self::Key, value: Self::Value) -> Option<Self::Value> {
+        // inserting an `EMPTY_VALUE` is equivalent to removing any value associated with `key`
+        if value != Self::EMPTY_VALUE {
+            self.perform_insert(key, value)
+        } else {
+            self.perform_remove(key)
+        }
+    }
+
+    fn get_leaf(&self, key: &Self::Key) -> Self::Leaf {
+        let leaf_pos = LeafIndex::<SMT_DEPTH>::from(*key).value();
+
+        match self.leaves.get(&leaf_pos) {
+            Some(leaf) => leaf.clone(),
+            None => SmtLeaf::Single((*key, Self::EMPTY_VALUE)),
+        }
+    }
+
+    fn hash_leaf(leaf: &Self::Leaf) -> RpoDigest {
+        leaf.hash()
+    }
+}
+
+impl Default for Smt {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+// SMT KEY
+// ================================================================================================
+
+/// Represents a key (256 bits) for the Smt.
+///
+/// The most significant `u64` determines the corresponding leaf index when inserting values into
+/// the Smt.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct SmtKey {
+    word: RpoDigest,
+}
```

Comment:
Do we need a newtype here? If it works, I think using `RpoDigest` directly would be more intuitive.

Reply:
The reason for the newtype was the [conversion to](https://github.com/0xPolygonMiden/crypto/blob/2179b86901d2060078ce6a6fdf5f0b221e8668b7/src/merkle/smt/mod.rs#L221-L226) `LeafIndex`. We could do `From<RpoDigest> to LeafIndex`, but then that would force all trees that use `RpoDigest` as a key to use this conversion implementation.

An alternative is to replace the `type Key: Into<LeafIndex>` trait bound in `SparseMerkleTree` with some required method `fn key_to_index(key) -> LeafIndex`.